clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -262,14 +293,66 @@ protected void onJoinComplete(int generation,
 
         // execute the user's callback after rebalance
         ConsumerRebalanceListener listener = subscriptions.rebalanceListener();
-        log.info("Setting newly assigned partitions: {}", Utils.join(assignedPartitions, ", "));
+
+        switch (protocol) {
+            case EAGER:
+                if (!ownedPartitions.isEmpty()) {
+                    log.info("Coordinator has owned partitions {} that are not revoked with {} protocol, " +
+                        "it is likely client is woken up before a previous pending rebalance completes its callback", ownedPartitions, protocol);
+                }
+
+                log.info("Setting newly assigned partitions: {}", Utils.join(assignedPartitions, ", "));
+                try {
+                    listener.onPartitionsAssigned(assignedPartitions);
+                } catch (WakeupException | InterruptException e) {
+                    throw e;
+                } catch (Exception e) {
+                    log.error("User provided listener {} failed on partition assignment", listener.getClass().getName(), e);
+                }
+                break;
+
+            case COOPERATIVE:
+                assignAndRevoke(listener, assignedPartitions, ownedPartitions);
+
+                if (assignment.error() == ConsumerProtocol.AssignmentError.NEED_REJOIN) {
+                    requestRejoin();
+                }
+
+                break;
+        }
+
+    }
+
+    private void assignAndRevoke(final ConsumerRebalanceListener listener,
```

Comment:
It would be helpful to have some high-level documentation somewhere in the code about the cooperative strategy and how it impacts . Perhaps this already exists? At a minimum, we should explain in the `ConsumerRebalanceListener` docs how the two strategies affect the API.

Reply:
Yeah my plan is to add the javadoc change in pt.3, and will do another doc PR for web-docs / upgrade-path; the rationale is that some of the behavior changes are yet to be agreed on in pt.3 so I'd like to defer that to later.